### PR TITLE
[SPARK-41171][SQL] Infer and push down window limit through window if partitionSpec is empty

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Sometimes, the SQL exists filter which condition compares the window function `ROW_NUMBER()` with number. For example,
```
SELECT *,
         ROW_NUMBER() OVER(ORDER BY a) AS rn
FROM Tab1
WHERE rn <= 5
```
We can create a `Limit` as the parent node of `Window` node. Such as: `Limit(Literal(5), Window)` and the optimizer rule `LimitPushDownThroughWindow` will push down the `Limit` as the child of `Window`. After this optimization, Spark executes top n and reduce the data size before shuffle. We can consider the SQL adjusted as the below.
```
SELECT *,
         ROW_NUMBER() OVER(ORDER BY a) AS rn
FROM 
    (SELECT *
    FROM Tab1
    ORDER BY  a LIMIT 5) t
```
In short, it supports following pattern:
```
SELECT (... row_number()
    OVER (
ORDER BY  ... ) AS rn)
WHERE rn (==|<|<=) k
        AND other conditions
```

### Why are the changes needed?
Improve the performance by infer window limit and push down it through window when partitionSpec is empty.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
The top n (`Limit` + `Sort`) have better performance than `WindowGroupLimit` if the window function is `RowNumber` and `Window`'s partitionSpec is empty.
Before this PR, the micro benchmark of Top-K show below.
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
Intel(R) Core(TM) i7-9750H CPU  2.60GHz
Benchmark Top-K:                                                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------
ROW_NUMBER (PARTITION: , WindowGroupLimit: false)                        10711          11068         363          2.0         510.8       1.0X
ROW_NUMBER (PARTITION: , WindowGroupLimit: true)                          2611           2929         210          8.0         124.5       4.1X
```
After this PR, the micro benchmark of Top-K show below.
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark Top-K:                                                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------
ROW_NUMBER (PARTITION: , WindowGroupLimit: false)                        11054          11684         406          1.9         527.1       1.0X
ROW_NUMBER (PARTITION: , WindowGroupLimit: true)                          1737           1772          23         12.1          82.8       6.4X
```
We can see the change of `ROW_NUMBER (PARTITION: , WindowGroupLimit: true) `: 2929 -> 1772.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan cc @wangyum

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan @dongjoon-hyun Thank you very much!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -130,6 +130,7 @@ abstract class Optimizer(catalogManager: CatalogManager)
     val operatorOptimizationBatch: Seq[Batch] = {
       Batch("Operator Optimization before Inferring Filters", fixedPoint,
         operatorOptimizationRuleSet: _*) ::
+      Batch("Infer window group limit", Once, InferWindowGroupLimit) ::
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:133]: I think this is an unrelated change and is actually a regression. We may infer filters and it's better to run window limit optimization as late as possible. @beliefer can we revert this change about moving the rule?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:133]: Got it.
